custom/aaa/models.py

```diff
@@ -0,0 +1,272 @@
+"""Contains all models for the AAA Convergence Dashboard.
+
+Conventions followed in this file:
+    * Fields ending in _ranges represent a field's history from forms that are useful for filtering.
+      This is only used for fields that are required to filter for specific ranges (usually a date range).
+    * Fields ending in _history represent a field's entire history from forms.
+      They are saved with the format [[date, value], [date, value]].
+      They are not guaranteed to be ordered by date.
+      They are not intended to be queried on, their interpretation should happen in python.
+      These history fields are saved in a separate model to ensure that the main model stays
+      fast and small for filtering.
+    * Each model contains all parts of the location hierarchy.
+    * Each model should support a future world where it is being accessed in a
+      sharded database (CitusDB).
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+from django.contrib.postgres.fields import ArrayField, DateRangeField
+from django.db import models
+
+from corehq.apps.userreports.models import StaticDataSourceConfiguration, get_datasource_config
+from corehq.apps.userreports.util import get_table_name
+
+
+class LocationDenormalizedModel(models.Model):
```

Comment:
Not that I think it applies heavily to these models but just mentioning for possibly future work that optimising column order should be considered for large tables: https://blog.2ndquadrant.com/on-rocks-and-sand/ 

It might actually make more sense for the UCR tables.

Reply:
Agreed that there's probably more low hanging fruit than this, but up to 20% space savings does seem worth exploring eventually.

I had heard of struct padding in C but never quite made the connection to SQL.